### PR TITLE
Bernhard dev

### PR DESCRIPTION
**Pull request by Kait0**
This pull request contains the following changes:

* Fix multi-GPU seeding bug
* Add a feature to disable advantage filtering via the flag adv_filter_enabled
It is true by default, so no functional change.
* Add feature min_batch_size, it will duplicate the collected data until the min_batch_size is reached.
The purpose is to enable a fixed amount of valid data to precisely control update steps.
(Due to invalid agents the amount of data collected each PPO iteration is variable).
The feature is disabled by default.
* Fix bugs and change naming structure of saved models to enable automatic evaluation
* Add SLURM scripts for automatic training and eval
* Fix nuPlan eval bugs (went into infinite loop if not enough RAM available, the PR reduces RAM usage and forces a crash instead of an infinite loop).
* Add adam_weight_decay: 0.01 as explicit parameter. Default is from AdamW to keep behavior the same.
* Some Claude-authored unit tests

Fixing the multi-GPU seeding bug affects performance. The other changes do not, as the default remains unchanged. (Might send a separate PR for config defaults update).

Below are three ablations. The last one shows that fixing the seeding bug improves performance by 25% and is the best trained model with 3.0 so far.
image

### CODE REVIEW

File: scripts/kesai/4_train_1GPU_slurm.sh
Comment (eugenevinitsky) [inline scripts/kesai/4_train_1GPU_slurm.sh:1]: can we move these into a kesai folder?

Comment (Kait0) [reply scripts/kesai/4_train_1GPU_slurm.sh:1]: done

File: pufferlib/pufferl.py [hunk 1]
```diff
@@ -1274,13 +1316,24 @@ def __init__(self, args, load_id=None, resume="allow"):
...
+
     def upload_model(self, model_path):
         artifact = self.wandb.Artifact(self.run_id, type="model")
         artifact.add_file(model_path)
         self.wandb.run.log_artifact(artifact)
 
     def upload_config(self, config_yaml_path):
-        self.wandb.save(config_yaml_path, base_path=os.path.dirname(config_yaml_path), policy="now")
+        # no-op: wandb.init(config=...) covers this.
+        pass
```
Comment (vcharraut) [inline pufferlib/pufferl.py]: are we sure about this ? It was added recently

File: .gitignore [hunk 2]
```diff
@@ -221,3 +221,5 @@ CLAUDE.local.md
 episode_metrics
 
 /obs
+
+bernhard_*
```
Comment (vcharraut) [inline .gitignore]: to delete

File: pufferlib/vector.py
Comment (vcharraut) [inline pufferlib/vector.py:1]: Could you provide a little explanation here or on Slack about this changes ?

File: pufferlib/pufferl.py (hunk 1 above)
Comment (Kait0) [reply pufferlib/pufferl.py]: I wanted to remove all these duplicate config files that the code was producing. Storing one config file on disk is enough.
This particular one overwrote the main config file since it had the same name.
Do you know who added it and why?

File: .gitignore (hunk 2 above)
Comment (Kait0) [reply .gitignore]: I use it to ignore my personal file. Is a problem?

Comment (eugenevinitsky) [reply .gitignore]: This is fine to me

File: pufferlib/pufferl.py (hunk 1 above)
Comment (eugenevinitsky) [reply pufferlib/pufferl.py]: I added it, the config that was being uploaded to wandb did not have the full values we needed. I missed that this overwrite would happen

Comment (eugenevinitsky) [reply pufferlib/pufferl.py]: Pushing a fix that removes the duplicate but resolves the reason we wound up with it in the first place

File: .gitignore (hunk 2 above)
Comment (vcharraut) [reply .gitignore]: For that you can edit .git/info/exclude that works as a personal gitignore

File: pufferlib/pufferl.py (hunk 1 above)
Comment (Kait0) [reply pufferlib/pufferl.py]: thanks

File: .gitignore (hunk 2 above)
Comment (Kait0) [reply .gitignore]: done

File: pufferlib/vector.py
Comment (Kait0) [reply pufferlib/vector.py:1]: The problem was that if one of the workers dies, this process would be an endless loop waiting for them. (resulting in the whole code hanging without output).
Most of the code waits for the env and eventually gives up if an env doesn't reply. 

And then there is some code to free some unused memory from setup.

File: pufferlib/vector.py [hunk 3]
```diff
@@ -16,6 +17,26 @@
 MAIN = 5
 INFO = 6
 
+# Driver busy-wait tuning. Spins are cheapest when a result is already there, so
+# a wait only starts tracking itself after the fast path misses. It then spins
+# hot -- covering any normal step -- before falling back to a poll. The poll is
+# short because the sleep is not what costs: a liveness check is one waitpid per
+# worker (~1.3us each), so it runs on its own much slower cadence.
```
Comment (eugenevinitsky) [inline pufferlib/vector.py]: Could I ask that we rewrite this to not sound like Claude before merging. This is my own nitpick, but Claude-sounding text drives me off the wall

File: pufferlib/vector.py [hunk 4]
```diff
@@ -16,6 +17,26 @@
...
+LIVENESS_CHECK_SECONDS = 0.01
+
+
+def _return_freed_heap_to_os():
+    """glibc parks large frees in its arena, so RSS -- what cgroup limits and the
+    OOM killer count -- stays at peak until trimmed. Only worth calling after
+    freeing something the process will not allocate again. No-op off glibc."""
+    try:
+        ctypes.CDLL("libc.so.6").malloc_trim(0)
+    except (OSError, AttributeError):
```
Comment (eugenevinitsky) [inline pufferlib/vector.py]: please no passes on try except like this without at least a message. This becomes a potential silent failure

File: pufferlib/vector.py (hunk 3 above)
Comment (eugenevinitsky) [reply pufferlib/vector.py]: If this is not LLM text, my apologies

File: pufferlib/pufferl.py [hunk 5]
```diff
@@ -705,33 +731,45 @@ def _train_ppo_transition(self, losses, profile, epoch):
...
+
+            if torch.distributed.is_initialized() and torch.distributed.get_world_size() > 1:
+                # Filtering keeps a different count per rank, so trim to the global
+                # minimum to keep the number of minibatches synchronized
+                kept_tensor = torch.tensor([num_kept], device=device)
+                torch.distributed.all_reduce(kept_tensor, op=torch.distributed.ReduceOp.MIN)
+                min_num_kept = kept_tensor.item()
+                if num_kept > min_num_kept:
+                    if min_num_kept == 0:
+                        keep_idx = keep_idx[:0]
```
Comment (eugenevinitsky) [inline pufferlib/pufferl.py:754]: I'm not sure I understand how this condition is supposed to work

Comment (eugenevinitsky) [inline pufferlib/pufferl.py:754]: I think this can't happen? Or, if it happens I don't really understand what happens here

File: pufferlib/vector.py
Comment (Kait0) [reply pufferlib/vector.py:1]: its quite a minor issue and a lot of code. i removed it now

File: pufferlib/vector.py (hunk 3 above)
Comment (Kait0) [reply pufferlib/vector.py]: removed the claude slop

File: pufferlib/vector.py (hunk 4 above)
Comment (Kait0) [reply pufferlib/vector.py]: removed

File: pufferlib/pufferl.py (hunk 5 above)
Comment (Kait0) [reply pufferlib/pufferl.py:754]: It is code that was already there so I don't know.
It just shows up green because I wrapped the whole section in an if statement.

What this does is that if any rank has 0 elements after advantage filtering, it will empty out the tensors on all other ranks as well, leading to the epoch being skipped down the line.

Whether that can happen, not sure.

**Review by eugenevinitsky — APPROVED** [review]
If you can attach nightly runs to

**Comment by eugenevinitsky** [comment]
This touches training so if possible please run nightly-multi-long in between so we have a checkpoint for it. Also, please make sure tests pass first.

**Comment by Kait0** [comment]
Oh right, these were my own unit tests, not needed anymore since I removed the code.
Can run nightly-multi-long, but since 3.0 seems degraded right now, I would expect a bad result?
Should I wait until this is fixed?

**Comment by Kait0** [comment]
The golden smoke test failure is what we discussed about on slack.
It happens because the last dangling batch after advantage filtering has batch size 1 and causes NAN when std of advantage is calculated.
Maybe I skip the last partial batch already here.
